### PR TITLE
Widgetbars

### PR DESCRIPTION
**Pull request by elbeardmorez**
I wanted to create plugins which utilised space in the main window. In order to facilitate this I extended the paned widgets to create 'expanding panes' (xpaneds) which I've named 'widgetbars' (for want of a better name)

**This PR adds limited functionality, a 'base' widgetbars plugin for which other PRs will be based**

2018mar11 widgetbars

The only useful thing this widgetbars base plugin can do is switch the order of the main window panes.

2018mar11 widgetbars2

I caught up late with the #152 work, but it appears this work is highly compatible given the ease of porting my work to use Nick's UserInterfacePlugin mechanism. Indeed the intrusion in quodlibetwindow was near identical.

I'm hoping that the commit names and any commit messages serve as sufficient to explain their purpose given the context of this pull request, and of course the supporting pics (which I do believe are worth a 1000 words).

If you would like to see the end game, the [elbeardmorez/quodlibet/widgetbars_all](https://github.com/elbeardmorez/quodlibet/tree/widgetbars_all) repo contains everything..

Other widgetbars are more exciting:

**CoversWidgetBar**: append all art for playing song to scrollable art set..

widgetbars 5

2018mar13 coverswidgetbar prefs

**PluginsWidgetBar**: easily switch plugins on/off and activate plugin related commands

widgetbars 3

2018mar13 pluginswidgetbar prefs

2018mar13 pluginswidgetbar actions

2018mar13 pluginswidgetbar icons drag n drop

**EmbeddedArtWidgetBar**: add / remove embedded art

2018apr01 embeddedartwidgetbar

2018apr01 embeddedartwidgetbar prefs

Other 'related extras' include:
-Multiple covers open at a time

widgetbars

### CODE REVIEW

**Comment by elbeardmorez** [comment]
The failing test is (IMO) due to the flakiness of PyGTK as a whole (note 6/8 passes). Specifically, I've made changes to a the 'make_paned_sizes_equal' as I removed the 'R for Relative' specificness of the MultiRPaned (now MultiPaned). In testing non-relative panes sizing (for which WidgetBars XPaned' are based), the accuracy in pixels on performing expand/collapses across multiple panes appears to differ depending on the wind and the number of clouds in the sky worldwide at any one time. I have accounted for that by adding a 'relatively_close' helper function which by default passes results within 1% of eachother. I didn't have failures my end. I imagine simply increasing this tolerance to 2% would 'work', but I personally don't want to. Your call! :)

I would also note that original tests ..i.e. the relative tests, also failed for a similar reason when they were applied to an actually <em>shown</em> window, hence the additional 'to 2 d.p.' passed to those tests.

File: quodlibet/quodlibet/ext/ui/widgetbars.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+class Config(object):
+    _config = PluginConfig(plugin_id)
+
+    pane_order = ConfProp(_config, "pane_order", "")
+
+
+CONFIG = Config()
+
+
+class WidgetBars(EventPlugin):
```
[the remark this answers is not in the document]
Comment (declension) [reply quodlibet/quodlibet/ext/ui/widgetbars.py]: Or perhaps this is more an _Advanced Preferences_ thing actually

File: quodlibet/quodlibet/_main.py [hunk 2]
```diff
@@ -19,7 +19,7 @@
 
 
 PLUGIN_DIRS = ["editing", "events", "playorder", "songsmenu", "playlist",
-               "gstreamer", "covers", "query"]
+               "gstreamer", "covers", "query", "ui"]
```
Comment (declension) [inline quodlibet/quodlibet/_main.py]: So far I guess nothing can be _just_ a UI plugin - that's more of a mixin than a full type - they nearly always end up being `EventPlugin`s too...

File: quodlibet/quodlibet/plugins/gui.py [hunk 3]
```diff
@@ -29,31 +31,55 @@ def create_sidebar(self):
...
-        sidebar = pl_obj.create_sidebar()
-        app.window.hide_side_book()
-        if sidebar:
-            print_d("Enabling sidebar for %s" % plugin.cls)
-            self.__sidebars[plugin] = app.window.add_sidebar(
-                sidebar, name=plugin.name)
-            sidebar.show_all()
+
+        # sidebars
+        if hasattr(pl_obj, 'create_sidebar'):
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/plugins/gui.py]: ah yes of course, they'll both get called regardless of what the subclasses actually implement, my mistake

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 4]
```diff
@@ -298,9 +301,14 @@ def __select_song(self, widget, indices, col, player):
 
 
 class TopBar(Gtk.Toolbar):
-    def __init__(self, parent, player, library):
+    def __init__(self, id, parent, player, library):
         super(TopBar, self).__init__()
 
+        self.id = id
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]: the id of the 'widget' ..akin to a plugin id, but a necessary addition to allow the topbar ('playbar') and the main browser ('main') to be added to the same dict container as the widgetbars

File: quodlibet/quodlibet/qltk/paned.py [hunk 5]
```diff
@@ -296,9 +525,33 @@ def __changed(self, widget=None, event=None):
...
+    def size_allocate(self, allocation):
+
+        alloc_size = allocation.height \
+            if self.orientation == Gtk.Orientation.VERTICAL \
+            else allocation.width
+
+        if alloc_size > self.default_size:
+            self.size = alloc_size
+
+#            print_d("%s | size_allocate event. storing height: %d" %
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/paned.py]: specfically the debug comment? if so, yes will do, i stripped the vast majority out, leaving only the few that i thought could be useful for anyone wanting to see 'what's going on' a little easier than starting from nothing. should i remove ALL commented debug lines?

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 6]
```diff
@@ -744,28 +765,36 @@ def on_queue_visible(qex, param):
         self.playlist = PlaylistMux(
             player, self.qexpander.model, self.songlist.model)
 
-        top_bar = TopBar(self, player, library)
-        main_box.pack_start(top_bar, False, True, 0)
+        # ui elements set
+        self.uielements = {}
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]: yep ok

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 7]
```diff
@@ -744,28 +765,36 @@ def on_queue_visible(qex, param):
...
-        main_box.pack_start(paned, True, True, 0)
+        # box for 'dynamic' paneds
+        self.__dynelements_box = Align()
+        main_box.pack_start(self.__dynelements_box, True, True, 4)
+
+        # browser / sidebar
+        self.__browserbox = Gtk.VBox()
+        self.__browserpaned = ConfigRHPaned("memory", "sidebar_pos", 0.25)
+        self.__browserpaned.pack1(self.__browserbox, resize=True)
+        # pack2 when the first sidebar plugin is set up
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]: this is the comment that it replaced..

-# We'll pack2 when necessary (when the first sidebar plugin is set up)

this, to me, is identical, just slightly more parsimonious

File: quodlibet/quodlibet/qltk/widgetbar.py [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+            self.panelock.updating = True
+            self.get_parent().update(self)
+            self.get_parent().check_resize()
+            self.panelock.updating = False
+
+    def __disable(self):
+        pm = PluginManager.instance
+        plugin = next((p for p in pm.plugins if p.id == self.id), None)
+        if plugin:
+            pm.enable(plugin, False)
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/widgetbar.py]: i think that that's what it does currently, the code is only disabling the single plugin. if that plugin happens to be the actual 'widgetbars' plugin, then just as you say, the Xwidgetbarplugins with still remain enabled (but not visible), only the widgetbars plugin is toggled

File: quodlibet/quodlibet/plugins/events.py [hunk 9]
```diff
@@ -59,6 +61,9 @@ def plugin_on_songs_selected(self, songs):
         """Called when the selection in main songlist changes"""
         pass
 
+    def plugin_on_plugin_toggled(self, plugin, enabled):
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/plugins/events.py]: see below

File: quodlibet/quodlibet/plugins/events.py [hunk 10]
```diff
@@ -97,7 +102,8 @@ def _map_signals(obj, prefix="plugin_on_", blacklist=None):
 
 class EventPluginHandler(PluginHandler):
 
-    def __init__(self, librarian=None, player=None, songlist=None):
+    def __init__(self, librarian=None, player=None,
+                 songlist=None, pluginmanager=None):
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/plugins/events.py]: i believe that this particular part is more applicable to the 'PluginsWidgetBarPlugin' commit set. i will have tried to strip it out, but failed as it's used by the 'Widgetbars Plugin' to update the main ui after it's disabled. the PluginsWidgetBarPlugin uses this ensure its knowledge of what's enabled or not remains in sync with the (previously solely responsible) PluginManager

File: quodlibet/quodlibet/qltk/widgetbar.py [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+from quodlibet.qltk import get_top_parent
+from quodlibet.qltk import Icons
+from quodlibet.qltk.menubutton import SmallMenuButton
+from quodlibet.qltk.x import ScrolledWindow, SymbolicIconImage, \
+    SmallImageButton, MenuItem, PaneLock
+
+from gi.repository import Gtk
+
+
+class WidgetBar(Gtk.Expander):
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/widgetbar.py]: :) yes of course, i will. i did mention that the name 'widgetbar' is not dear to me too so if you wanted to just do a %s/widgetbars/new name/g globally if/when this goes in, i wouldn't be offended!

File: quodlibet/quodlibet/qltk/paned.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (declension) [reply quodlibet/quodlibet/qltk/paned.py]: Well a few aren't a problem, but in general best to either make it a helpful comment for the reader, or an actual debug message for the user / developer

File: quodlibet/quodlibet/qltk/quodlibetwindow.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (declension) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]: heh, think that was me actually...

File: quodlibet/quodlibet/qltk/widgetbar.py [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+        self.panelock.default_size = value
+
+    def preferences(self, data):
+        if self.preferences_cb:
+            self.preferences_cb()
+
+    def __preferences(self):
+        plugins_window = PluginWindow(get_top_parent(self))
+        plugins_window.show()
+        if id:
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/widgetbar.py]: yes it is, thanks! the if statement is no-op so functionally it's as expected. the function is a broken cross between the an original implementation and the overridden __preferences in the PluginsWidgetBarPlugin. good spot

File: quodlibet/quodlibet/qltk/paned.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/paned.py]: agree totally! the issue i had here was that most of the comments i wanted to leave in place for other devs, are in places that would flood the terminal if left in, rendering --debug mode useless for any other purpose eg. in do_size_allocate

File: quodlibet/quodlibet/ext/ui/widgetbars.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/ext/ui/widgetbars.py]: yes, i had no clue where to put it when i made that decision initially. let's just stick to what's there, as you say, it's an events plugin ..put it in /events

File: quodlibet/quodlibet/ext/ui/widgetbars.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    """The plugin class."""
+
+    PLUGIN_ID = "widgetbars"
+    PLUGIN_NAME = _("WidgetBars")
+    PLUGIN_DESC = _("Control plugin for WidgetBars.")
+    PLUGIN_CONFIG_SECTION = __name__
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/ext/ui/widgetbars.py]: hmm, why so? do you mean just in that plugin, or across the board? i've certainly used these ID and NAME values in one of the plugins

File: quodlibet/quodlibet/_main.py (hunk 2 above)
Comment (elbeardmorez) [reply quodlibet/quodlibet/_main.py]: i'll ditch 'ui' then, although it was more a chosen folder name for the 'PluginsWidgetBarPlugin', which to me 'works', but then as say it is just 'Advanced Preferences', and it's an eventplugin so ../events

File: quodlibet/quodlibet/qltk/paned.py [hunk 14]
```diff
@@ -119,6 +120,210 @@ class RVPaned(RPaned):
...
+        """Pass the widget whose pane needs updating."""
+
+        self.__updating = True
+
+        if not isinstance(widget, Gtk.Expander):
+            return
+
+        # lock other non-expander or expanded widgets
+        widgets = self.panelocks(self.root)
+        for w in (w for w in widgets
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/paned.py]: i'm just scratching my head looking at that stupidity myself (it's the two lines below that too right?!), yes i'll change it

File: quodlibet/quodlibet/ext/ui/widgetbars.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (declension) [reply quodlibet/quodlibet/ext/ui/widgetbars.py]: no just the `PLUGIN_CONFIG_SECTION`. It doesn't really do anything AIUI. Yes, there are a couple of others that seem to do this too actually.

However if you define a `CONFIG_SECTION` and inherit from `PluginConfigMixin` (although that's kinda deprecated: #1827) then that will work and if not it munged the ID.

File: quodlibet/quodlibet/plugins/events.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (declension) [reply quodlibet/quodlibet/plugins/events.py]: Hmm I see

File: quodlibet/quodlibet/qltk/paned.py [hunk 15]
```diff
@@ -119,6 +120,210 @@ class RVPaned(RPaned):
...
+            p = p.get_parent()
+            p.check_resize()
+
+        self.__updating = False
+
+    def do_size_allocate(self, *args):
+        # clear size requests to allow shrinking
+        Gtk.HPaned.do_size_allocate(self, *args)
+
+        if not self.__updating:
```
Comment (declension) [inline quodlibet/quodlibet/qltk/paned.py]: BTW indentation looks neater with a `if self.__updating: return` pattern (as below in fact)

File: quodlibet/quodlibet/qltk/paned.py [hunk 16]
```diff
@@ -119,6 +120,210 @@ class RVPaned(RPaned):
...
+        self.__updating = True
+
+        if not isinstance(widget, Gtk.Expander):
+            return
+
+        # lock other non-expander or expanded widgets
+        widgets = self.panelocks(self.root)
+        for w in (w for w in widgets
+            if w is not widget
+                 and (not isinstance(w, Gtk.Expander) or w.get_expanded())):
```
Comment (declension) [inline quodlibet/quodlibet/qltk/paned.py]: So long as we (this class) controls all the packing, I think we should be able to omit the various instance checks, they make me OOP-nervous anyway

Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/paned.py]: the trouble is, we need to consider other non-expander types that are in the mix too! again, 'main' (now a ConfigRHPaned obj) and 'playbar' (TopBar obj) are parts of resizing equation. it did look nice and clean before i started working with two or three nested paneds, and that's when all heck broke loose. that chunk of code is now very simple to look though.

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 17]
```diff
@@ -834,6 +861,98 @@ def on_queue_visible(qex, param):
...
+                    if key not in order]
+            order.extend(extra)
+
+            self.uielements_ordered.extend(
+                [val for key, val in self.uielements.items()
+                    if key in extra])
+
+        print_d("%d ui elements live:" % len(self.uielements_ordered))
+        for w in self.uielements_ordered:
+            print_d("%r" % w)
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]: very good point

File: quodlibet/quodlibet/qltk/quodlibetwindow.py [hunk 18]
```diff
@@ -834,6 +861,98 @@ def on_queue_visible(qex, param):
...
+            self.__dynelements_box.remove(
+                self.__dynelements_box.get_children()[0])
+
+        # add ui elements to multipane
+        print_d("adding %d widgets to multipane" % len(uielements))
+        self.__multipaned.set_widgets(uielements, [(False, False)])
+
+        # ensure initial panelocks
+        for w in uielements:
+            if hasattr(w, 'panelock') and not (isinstance(w, Gtk.Expander)
```
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]:
```
 883         # ensure initial panelocks$
 884         def lockable(widget):$
 885             return hasattr(widget, 'panelock') and$
 886                 not (isinstance(w, Gtk.Expander) and not w.get_expanded())$
 887 $
 888         for w in ui_elements:$
 889             if not lockable(w):$
 890                 continue$

```
e.g. ?? better?

[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/quodlibetwindow.py]: cool, thanks

File: quodlibet/quodlibet/qltk/paned.py (hunk 15 above)
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/paned.py]: yeh, like 10 lines below, that's hilarious (it's actually tragic/painful ..i'm clueless as to why i'd think both ways). changed.

File: quodlibet/quodlibet/qltk/widgetbar.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (elbeardmorez) [reply quodlibet/quodlibet/qltk/widgetbar.py]: see [commit] (https://github.com/quodlibet/quodlibet/pull/2552/commits/5f302d8a0a18df16fbe33e8b62bb0a7df92830e5). comment is below the above code hence no 'outdated' trigger

**Comment by declension** [comment]
@elbeardmorez BTW you can use `self.failUnlessAlmostEqual(actual, expected, delta=123)` (see documentation) instead of writing your own, and this fails more helpfully than `AssertionError: False is not true` as per CI builds above...

**Comment by elbeardmorez** [comment]
@declension thanks, i think they're a little different though. i'd looked at using that function (and indeed i made correct use of it in the lines above where i had to change the existing test to work under a 'live' window) ..before creating my own. it may sound the same, but it doesn't do the same thing. 'relatively close' for me here needed to be relative to the size of the container, and simply testing two numbers for their difference at a specific order (decimal point position) isn't what i wanted

point definitely taken about the lack of useful output though, i've pushed a change which should help

**Comment by elbeardmorez** [comment]
there's something radically different with that ubuntu world ..the diffs are 24% vs the 1% tolerance passes for the other boxes/distros ..any ideas?

**Comment by declension** [comment]
> there's something radically different with that ubuntu world ..the diffs are 24% vs the 1% tolerance passes for the other boxes/distros ..any ideas?

Great, that's a much clearer problem now. I thought it might be something a bit more drastic, though unfortunately I have no clue why Ubuntu 16.04 behaves so differently. I guess it's a different GTK+ version, but still.

Have you tried _using_ it on 16.04? Perhaps that's the next step?

**Comment by elbeardmorez** [comment]
@declension: i'm there now, and the truth is ugly :(. although the test only fails at 3.1% (129px vs 125px), there's an issue with the resizing that has me back in paned.py (my least favourite place in the world). on the plus side however, i've found several other bugs, py3 and non-py3 related that needed fixing that i otherwise would have missed!

**Comment by elbeardmorez** [comment]
well here's part of the 'special' patch that caused the above scare..

```@@ -253,26 +253,27 @@ class XPaned(Paned):
 52          # clear size requests to allow shrinking
 53          Gtk.HPaned.do_size_allocate(self, *args)
 54
 55 -        if not self.__updating:
 56 +        if self.__resizing:
 57 +            return
 58
 59 -            if not self.__panelocks:
 60 -                widgets = self.panelocks(self.root)
`GB
```

..note the typo _'__updating'_ != _'__resizing'! that has been (shamelessly) overwritten now <- it **never** happened..

**Comment by elbeardmorez** [comment]
the Gtk version difference and our loading of different css data based on that was the ultimate difference in Debian vs Ubuntu (pass vs fail) in this patch set

the [paned handle position considerations](https://github.com/quodlibet/quodlibet/pull/2552/commits/4a709422382dfde796a620ab3e5fabdb21d5f269) patch is what has finally allowed this work to pass on those boxes ..having worked with GtkInspector, and a dummy Gtk app on a Ubuntu 17 virtual box 

the additional comments in that patch are worth looking at! also, i must mention again that the original tests, and indeed all functionality around _'make_paned_sizes_equal'_ were not only inaccurate when a handle size > 1 px was used, but were totally useless given that no **live**  window was used. we were effectively saying 'given a size allocated (always 1!), lets set the handle position to 1/4 ..now test if get_relative (e.g. 1/4) of 1 is 1/4, True? great, pass!

the tests now calculate the expected size of a pane and check each pane size against that as opposed to where the position of the handle is in that paned.

to do this functionality correctly, the size of the handle must be known! we're now straying into painful territory with the advent of loading css data pitted again possibly contrary user css files. although the tests for this commit set **would** pass without the last commit [add a 'force' option to the 'css_add' wrapper](https://github.com/quodlibet/quodlibet/pull/2552/commits/3079913cbb433b9960ce5cfa23f3817332b53c94), i've added it to ensure the functionality still works for all users who happen to keep custom USER priority gtk css files (one of which is yours truly!)

example of breakage: consider a window size 200 containing a single paned, with our handle size 6px (pre-commit). the original _'make_paned_sizes_equal'_ said, _'set_position(100)'_ which resulted in pane 1 at 100px, pane2 at 94px thus not equal, but alas, even when setup as a live window and 100 should have been compared to 94 and failed, we still did the test 'is _get_position_  == 1/2 of 200', and yes 100 == 100, pass! now also consider a user css file overriding our 6px handle size with 20px ..and that issue was compounded!

**Comment by declension** [comment]
Closing old MRs (#3655)

**Pull request closed without merging**